charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,252 @@
+        self._k8s_resource_handler = None
 
-            self._check_leader()
-
-            interfaces = self._get_interfaces()
-
-            image_details = self._check_image_details()
+        http_port = ServicePort(int(self._http_port), name="http")
+        self.service_patcher = KubernetesServicePatch(
+            self, [http_port], service_name=f"{self._name}"
+        )
 
-        except CheckFailed as check_failed:
-            self.model.unit.status = check_failed.status
-            return
+        # Set up event handlers
+        self.framework.observe(self.on.upgrade_charm, self._on_event)
+        self.framework.observe(self.on.config_changed, self._on_event)
+        self.framework.observe(self.on.leader_elected, self._on_event)
+        self.framework.observe(self.on["ingress"].relation_changed, self._on_event)
+        self.framework.observe(self.on.tensorboards_web_app_pebble_ready, self._on_event)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.remove, self._on_remove)
+
+    @property
+    def container(self) -> Container:
+        """Return tensorboard-controller container object."""
+        return self._container
+
+    @property
+    def k8s_resource_handler(self) -> KubernetesResourceHandler:
+        """Get the K8S resource handler."""
+        context = {
+            "app_name": self._name,
+            "namespace": self._namespace,
+            "service": self._name,
+        }
+
+        if not self._k8s_resource_handler:
+            self._k8s_resource_handler = KubernetesResourceHandler(
+                field_manager=LIGHTKUBE_FIELD_MANAGER,
+                template_files=K8S_RESOURCE_FILES,
+                context=context,
+                logger=self.logger,
+            )
+        load_in_cluster_generic_resources(self._k8s_resource_handler.lightkube_client)
+        return self._k8s_resource_handler
 
-        self._configure_mesh(interfaces)
+    @k8s_resource_handler.setter
+    def k8s_resource_handler(self, handler: KubernetesResourceHandler):
+        """Set the K8S resource handler"""
+        self._k8s_resource_handler = handler
 
+    @property
+    def _env_vars(self) -> Dict[str, str]:
+        """Return environment variables based on model configuration."""
         config = self.model.config
+        env_vars = {
+            "APP_PREFIX": "/tensorboards",
+            "APP_SECURE_COOKIES": str(config["secure-cookies"]),
+            "BACKEND_MODE": config["backend-mode"],
+            "USERID_HEADER": "kubeflow-userid",
+            "USERID_PREFIX": "",
+        }
+
+        return env_vars
+
+    @property
+    def _tensorboards_web_app_layer(self) -> Layer:
+        """Create and return Pebble framework layer."""
+        exec_command = (
+            "gunicorn"
+            " -w 3"
+            f" --bind 0.0.0.0:{self._http_port}"
+            " --access-logfile"
+            " - entrypoint:app"
+        )
```

Comment:
Indeed, many of our charms use this pattern. I decided to move this away as part of the app's layer since we 're only using it there and it felt confusing when reading the code that we introduced all these attributes with no context as to their pupose of use. I don't have a strong preference on this though, I 'm ok with moving it back for consistency purposes if you think we should.

Reply:
That's alright, you can keep it as is!